sklearn/utils/validation.py

```diff
@@ -239,19 +240,30 @@ def _ensure_sparse_format(spmatrix, accept_sparse, dtype, copy,
     spmatrix_converted : scipy sparse matrix.
         Matrix that is ensured to have an allowed type.
     """
-    if accept_sparse in [None, False]:
-        raise TypeError('A sparse matrix was passed, but dense '
-                        'data is required. Use X.toarray() to '
-                        'convert to a dense numpy array.')
     if dtype is None:
         dtype = spmatrix.dtype
 
     changed_format = False
-    if (isinstance(accept_sparse, (list, tuple))
-            and spmatrix.format not in accept_sparse):
-        # create new with correct sparse
-        spmatrix = spmatrix.asformat(accept_sparse[0])
-        changed_format = True
+
+    if isinstance(accept_sparse, six.string_types):
+        accept_sparse = [accept_sparse]
+
+    if accept_sparse is False:
+        raise TypeError('A sparse matrix was passed, but dense '
+                        'data is required. Use X.toarray() to '
+                        'convert to a dense numpy array.')
+    elif (isinstance(accept_sparse, (list, tuple)) and
+          len(accept_sparse)):
```

Comment:
Hmmm so what is the behaviour of `accept_sparse=[]`, Is it the same as `accept_sparse=False`? Maybe this should be tackled inside this if clause to be more clear, i.e. something like:

```py
elif ininstance(accept_sparse, (list, tuple)):
    if len(accept_sparse) == 0:
        raise ValueError("When providing 'accept_sparse' as a tuple or list it should be non empty or some better errror message")
```